creator-node/src/middlewares.js

```diff
@@ -56,47 +72,78 @@ async function syncLockMiddleware (req, res, next) {
  * Blocks writes if node is not the primary for audiusUser associated with wallet
  */
 async function ensurePrimaryMiddleware (req, res, next) {
-  const serviceRegistry = req.app.get('serviceRegistry')
   const start = Date.now()
+  let logPrefix = '[ensurePrimaryMiddleware]'
+  const logger = req.logger
+
+  const serviceRegistry = req.app.get('serviceRegistry')
+  const { nodeConfig, libs } = serviceRegistry
 
   if (!req.session || !req.session.wallet) {
-    return sendResponse(req, res, errorResponseUnauthorized('User must be logged in'))
+    return sendResponse(req, res, errorResponseUnauthorized(`${logPrefix} User must be logged in`))
   }
 
-  let serviceEndpoint
-  try {
-    serviceEndpoint = await getOwnEndpoint(serviceRegistry)
-  } catch (e) {
-    return sendResponse(req, res, errorResponseServerError(e))
+  if (!req.session.userId) {
+    return sendResponse(req, res, errorResponseBadRequest(`${logPrefix} User must specify 'User-Id' request header`))
   }
+  const userId = req.session.userId
 
-  let creatorNodeEndpoints
+  logPrefix = `${logPrefix} [userId = ${userId}]`
+
+  const selfSpID = nodeConfig.get('spID')
+  if (!selfSpID) {
+    return sendResponse(req, res, errorResponseServerError(`${logPrefix} Node failed to recover its own spID. Cannot validate user write.`))
+  }
+
+  /**
+   * Fetch current user replicaSetSpIDs
+   * Will throw error if selfSpID is not user primary
+   */
+  let replicaSetSpIDs
   try {
-    creatorNodeEndpoints = await getCreatorNodeEndpoints({
+    replicaSetSpIDs = await getReplicaSetSpIDs({
       serviceRegistry,
       logger: req.logger,
       wallet: req.session.wallet,
       blockNumber: req.body.blockNumber,
       ensurePrimary: true,
-      myCnodeEndpoint: serviceEndpoint
+      selfSpID,
+      userId
     })
   } catch (e) {
-    return sendResponse(req, res, errorResponseServerError(e))
+    return sendResponse(req, res, errorResponseServerError(`${logPrefix} ${e.message}`))
   }
-  const primary = creatorNodeEndpoints[0]
-
-  // Error if this node is not primary for user.
-  if (!primary || (primary && serviceEndpoint !== primary)) {
-    return sendResponse(
-      req,
-      res,
-      errorResponseUnauthorized(`This node (${serviceEndpoint}) is not primary for user. Primary is: ${primary}`)
-    )
+
+  const primarySpID = replicaSetSpIDs[0]
+
+  // TODO there should be no need to do this validation (standardize / cleanup errors between here and in getReplicaSetSpIDs())
```

Comment:
its just duplicated - the inner function already will throw on this condition (line 527) which means that this line here is never hit - instead the error is caught and thrown on line 116 as a serverError instead of Unauthorized - i didn't change it just in favor of minimizing scope of changes. all this logic is messy but i didn't want to change too much at once